### PR TITLE
[All Packages] Pass keys in all packages

### PR DESCRIPTION
**Pull request by iketw**
-  Eliminate ThirdwebStorage empty creation in SDK
- Create new getValidChainRPCs in the chains package
- Pass clientId in react-core, react, react-native and unity-bridge
- Pass clientId and secretKey in: wallets, sdk, storage
- Delete all default api keys
- Warn devs when not passing a clientId - they will get limited access to our services
- Rename paper's clientId to paperClientId

### CODE REVIEW

File: packages/storage/src/core/storage.ts [hunk 1]
```diff
@@ -24,8 +24,8 @@ import { IpfsUploader } from "./uploaders/ipfs-uploader";
  *
  * @example
  * ```jsx
- * // Create a default storage class without any configuration
- * const storage = new ThirdwebStorage();
+ * // Create a default storage class with an API key
+ * const storage = new ThirdwebStorage({apiKey: "your-api-key"});
```
Comment (joaquim-verges) [inline packages/storage/src/core/storage.ts:28]: nit: space here and above

File: packages/sdk/src/evm/schema/sdk-options.ts [hunk 2]
```diff
@@ -11,7 +11,8 @@ export const SDKOptionsSchema = /* @__PURE__ */ (() =>
     .object({
       // @ts-expect-error - zod doesn't know anything about readonly
       supportedChains: z.array(ChainInfoInputSchema).default(defaultChains),
-      thirdwebApiKey: z.string().default(DEFAULT_API_KEY),
+      apiKey: z.string().optional(),
+      thirdwebApiKey: z.string().default(DEFAULT_API_KEY), // DEPRECATED: Use `apiKey` instead
       alchemyApiKey: z.string().optional(),
       infuraApiKey: z.string().optional(),
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/schema/sdk-options.ts:17]: very tempted to get rid of these extra apiKeys

File: packages/wallets/src/evm/connectors/smart-wallet/index.ts [hunk 3]
```diff
@@ -32,13 +31,17 @@ export class SmartWalletConnector extends Connector<SmartWalletConnectionArgs> {
 
   constructor(config: SmartWalletConfig) {
     super();
+    if (!config.apiKey && config.thirdwebApiKey) {
```
Comment (joaquim-verges) [inline packages/wallets/src/evm/connectors/smart-wallet/index.ts:34]: ugh this is annoying because thirdwebApiKey is mandatory here... so even if you want to get rid of it you cant.

actually let's make `thirdwebApiKey` optional as well here, it technically makes sense since you can pass your own bundlerURL if you want to

File: packages/wallets/src/evm/connectors/smart-wallet/types.ts [hunk 4]
```diff
@@ -18,7 +18,11 @@ import { WalletConnectReceiverConfig } from "../../../core/types/walletConnect";
 export type SmartWalletConfig = {
   chain: ChainOrRpcUrl;
   factoryAddress: string;
+  /**
+   * @deprecated Use `apiKey` instead
+   */
   thirdwebApiKey: string;
```
Comment (joaquim-verges) [inline packages/wallets/src/evm/connectors/smart-wallet/types.ts:24]: make this optional

File: packages/react-native/src/evm/providers/thirdweb-provider.tsx [hunk 5]
```diff
@@ -76,7 +74,6 @@ export const ThirdwebProvider = <
 
   return (
     <ThirdwebProviderCore
-      thirdwebApiKey={thirdwebApiKey}
```
Comment (MananTank) [inline packages/react-native/src/evm/providers/thirdweb-provider.tsx:79]: @iketw Why is this removed? shouldn't it be passed along to ThirdwebProviderCore ?

File: packages/sdk/src/evm/constants/urls.ts [hunk 6]
```diff
@@ -61,9 +61,7 @@ export function getChainProvider(
 
   // if we still don't have an url fall back to just using the chainId or slug in the rpc and try that
   if (!rpcUrl) {
-    rpcUrl = `https://${chainId || network}.rpc.thirdweb.com/${
-      options.thirdwebApiKey || DEFAULT_API_KEY
-    }`;
+    rpcUrl = `https://${chainId || network}.rpc.thirdweb.com/${options.apiKey}`;
```
Comment (MananTank) [inline packages/sdk/src/evm/constants/urls.ts:64]: Shouldn't we fallback `apiKey` to `thirdwebApiKey` here? - `${options.apiKey || options.thirdwebApiKey}`

File: packages/sdk/src/evm/schema/sdk-options.ts [hunk 7]
```diff
@@ -11,7 +11,8 @@ export const SDKOptionsSchema = /* @__PURE__ */ (() =>
     .object({
       // @ts-expect-error - zod doesn't know anything about readonly
       supportedChains: z.array(ChainInfoInputSchema).default(defaultChains),
-      thirdwebApiKey: z.string().default(DEFAULT_API_KEY),
+      apiKey: z.string().optional(),
+      thirdwebApiKey: z.string().default(DEFAULT_API_KEY), // DEPRECATED: Use `apiKey` instead
```
Comment (MananTank) [inline packages/sdk/src/evm/schema/sdk-options.ts:15]: Do we have to use `.default()` here? - can we not make it `.optional()` ?

**Comment by MananTank** [comment]
When passing neither `apiKey` or `thirdwebApiKey` in React - I'm getting below warning which is confusing. 

This is happening because `thirdwebApiKey` is not undefined and is the `DEFAULT_API_KEY` because of sdk options schema

image

**Comment by MananTank** [comment]
There are a few places where `ThirdwebStorage` is initialized without apiKey

* sdk/src/evm/core/classes/transactions.ts
* sdk/src/evm/functions/utils/cache.ts
* sdk/src/solana/sdk.ts
* sdk/src/solana/server/index.ts

File: packages/react-native/src/evm/providers/thirdweb-provider.tsx (hunk 5 above)
Comment (iketw) [reply packages/react-native/src/evm/providers/thirdweb-provider.tsx:79]: it is being passed down as part of `{...restProps}`, the only reason we specified it here was because we were providing a default api key

File: packages/sdk/src/evm/constants/urls.ts (hunk 6 above)
Comment (iketw) [reply packages/sdk/src/evm/constants/urls.ts:64]: good catch!

**Comment by iketw** [comment]
@MananTank can your PR: https://github.com/thirdweb-dev/js/pull/1319 on top of this one?

File: packages/react-native/src/evm/providers/thirdweb-provider.tsx (hunk 5 above)
Comment (MananTank) [reply packages/react-native/src/evm/providers/thirdweb-provider.tsx:79]: @iketw restProps does not include `thirdwebApiKey`

Comment (iketw) [reply packages/react-native/src/evm/providers/thirdweb-provider.tsx:79]: what do you mean? Because we removed thirdwebApiKey from the destructure here:

```
{
  children,
  createWalletStorage = createAsyncLocalStorage,
  supportedWallets = DEFAULT_WALLETS,
  authConfig,
  theme,
  ...restProps
}: PropsWithChildren<ThirdwebProviderProps<TChains>>
```

now `restProps` has the rest of the props, you can try doing `restProps.thirdwebApiKey`

File: packages/wallets/src/evm/connectors/smart-wallet/types.ts [hunk 8]
```diff
@@ -18,7 +18,7 @@ import { WalletConnectReceiverConfig } from "../../../core/types/walletConnect";
 export type SmartWalletConfig = {
   chain: ChainOrRpcUrl;
   factoryAddress: string;
-  thirdwebApiKey: string;
+  thirdwebApiKey?: string;
```
Comment (MananTank) [inline packages/wallets/src/evm/connectors/smart-wallet/types.ts:21]: Let's make this mandatory again?

**Comment by MananTank** [comment]
What are we thinking about the RPC - if no apiKey is provided then we use the default API key?

File: packages/wallets/src/evm/connectors/smart-wallet/types.ts (hunk 8 above)
Comment (joaquim-verges) [reply packages/wallets/src/evm/connectors/smart-wallet/types.ts:21]: actually let's leave it optional, makes more sense when you're using your own bundler to not pass it

File: packages/wallets/src/evm/wallets/abstract.ts [hunk 9]
```diff
@@ -46,11 +47,29 @@ export async function checkContractWalletSignature(
...
 {
   public type: Ecosystem = "evm";
+  private params: AbstractWalletParams;
+
+  constructor(params: AbstractWalletParams) {
+    super();
+
+    if (!params.apiKey) {
+      console.warn(
+        "No API key provided. You will have limited access to thirdweb's services for storage, RPC, and account abstraction. You can get an API key from https://thirdweb.com/dashboard/",
```
Comment (joaquim-verges) [inline packages/wallets/src/evm/wallets/abstract.ts:66]: not going to warn here since it's not clear why you need it in wallets (yet)

removing this warn

File: packages/wallets/src/lib/wagmi-connectors/WagmiConnector.ts [hunk 10]
```diff
@@ -39,8 +48,17 @@ export abstract class WagmiConnector<
     options: Options;
   }) {
     super();
-    this.chains = chains;
+    if (!options.apiKey) {
+      console.warn(
+        "No API key provided. You will have limited access to thirdweb's services for storage, RPC, and account abstraction. You can get an API key from https://thirdweb.com/dashboard/",
```
Comment (joaquim-verges) [inline packages/wallets/src/lib/wagmi-connectors/WagmiConnector.ts:53]: same here

File: packages/sdk/src/evm/core/wallet/user-wallet.ts [hunk 11]
```diff
@@ -112,13 +119,19 @@ export class UserWallet {
...
-        ),
-      };
+      const transferTransaction = await Transaction.fromContractInfo({
+        contractAddress: resolvedCurrency,
+        contractAbi: ERC20Abi,
+        provider: this.connection.getProvider(),
+        signer,
+        method: "transfer",
+        args: [resolvedTo, amountInWei],
+        storage: new ThirdwebStorage({
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/core/wallet/user-wallet.ts:129]: ugh a bit annoying to make it mandatory here, @MananTank react-core level fix didnt work i assume?

since we're doing this might as well do the old flow with ContractWrapper tbh

File: packages/react-native/src/evm/providers/thirdweb-provider.tsx [hunk 12]
```diff
@@ -74,6 +74,13 @@ export const ThirdwebProvider = <
     [supportedWallets],
   );
 
+  if (!thirdwebApiKey) {
+    console.warn(
+      "No API key provided. You will have limited access to thirdweb's services for storage, RPC, and account abstraction. You can get an API key from https://thirdweb.com/dashboard/",
```
Comment (joaquim-verges) [inline packages/react-native/src/evm/providers/thirdweb-provider.tsx:79]: wait this will print on every re-render right? we should only print it once

File: packages/react-core/src/evm/providers/thirdweb-sdk-provider.tsx [hunk 13]
```diff
@@ -179,11 +179,17 @@ export const ThirdwebSDKProvider = <TChains extends Chain[]>({
...
   activeChain,
-  thirdwebApiKey = DEFAULT_API_KEY,
+  thirdwebApiKey,
   alchemyApiKey,
   infuraApiKey,
   ...restProps
 }: React.PropsWithChildren<ThirdwebSDKProviderProps<TChains>>) => {
+  if (!thirdwebApiKey) {
+    console.warn(
+      "No API key provided. You will have limited access to thirdweb's services for storage, RPC, and account abstraction. You can get an API key from https://thirdweb.com/dashboard/",
```
Comment (joaquim-verges) [inline packages/react-core/src/evm/providers/thirdweb-sdk-provider.tsx:189]: same here - we don't want to print it twice betwee ThirdwebProvider / ThirdwebSDKProvider

In fact i think we can just print it here, but we need to do a useRef or something to not print on every render

File: packages/sdk/src/evm/core/wallet/user-wallet.ts (hunk 11 above)
Comment (joaquim-verges) [reply packages/sdk/src/evm/core/wallet/user-wallet.ts:129]: reverted it

File: packages/wallets/src/lib/wagmi-connectors/WagmiConnector.ts (hunk 10 above)
Comment (joaquim-verges) [reply packages/wallets/src/lib/wagmi-connectors/WagmiConnector.ts]: removed

File: packages/wallets/src/evm/interfaces/connector.ts [hunk 14]
```diff
@@ -23,11 +32,12 @@ export type ConnectParams<TOpts extends Record<string, any> = {}> = {
 
 export class WagmiAdapter<
   TConnectParams extends Record<string, any> = {},
-> extends Connector<TConnectParams> {
+> extends Connector<TConstructorParams, TConnectParams> {
   wagmiConnector: WagmiConnector<any, any, any>;
 
   constructor(wagmiConnector: WagmiConnector) {
-    super();
+    super({ apiKey: "pepe" });
```
Comment (MananTank) [inline packages/wallets/src/evm/interfaces/connector.ts:39]: lol, what?

**Comment by MananTank** [comment]
Issues

* wallet's apiKey is not passed to connectors - connector is always using the default apiKey
* connectors' chains array is updated to use default apiKey but not the wallet
* Magic wallet already has a apiKey - what should we do about that? 
* should we pass down the apiKey to wallets from react-core too?

File: packages/react-core/src/evm/providers/thirdweb-sdk-provider.tsx (hunk 13 above)
Comment (iketw) [reply packages/react-core/src/evm/providers/thirdweb-sdk-provider.tsx]: Notice that this will only happen if the user users sdk directly, for react/rn, we catch this in those packages and pass a default so this will not trigger.

Will look at the re-rendering

File: packages/react-native/src/evm/providers/thirdweb-provider.tsx (hunk 12 above)
Comment (MananTank) [reply packages/react-native/src/evm/providers/thirdweb-provider.tsx]: fixed

File: packages/react-core/src/evm/providers/thirdweb-sdk-provider.tsx (hunk 13 above)
Comment (MananTank) [reply packages/react-core/src/evm/providers/thirdweb-sdk-provider.tsx]: fixed

File: packages/wallets/src/evm/connectors/smart-wallet/lib/http-rpc-client.ts [hunk 15]
```diff
@@ -25,7 +26,11 @@ export class HttpRpcClient {
       {
         url: this.bundlerUrl,
         headers: {
-          "x-api-key": apiKey,
+          ...(clientId
+            ? { "x-client-id": clientId }
+            : secretKey
+            ? { "x-secret-key": secretKey }
+            : {}),
```
Comment (MananTank) [inline packages/wallets/src/evm/connectors/smart-wallet/lib/http-rpc-client.ts:33]: same here - If `clientId` and `secretKey` both are given - only `x-client-id` header is set

File: packages/storage/src/core/uploaders/ipfs-uploader.ts [hunk 16]
```diff
@@ -112,11 +112,15 @@ export class IpfsUploader implements IStorageUploader<IpfsUploadBatchOptions> {
           : "Storage SDK",
     };
 
-    if (this.secretKey) {
-      headers["x-secret-key"] = this.secretKey;
+    if (this.secretKey && this.clientId) {
+      throw new Error(
+        "Cannot use both secret key and client ID. Please use secretKey for server-side applications and clientId for client-side applications.",
+      );
     }
```
Comment (MananTank) [inline packages/storage/src/core/uploaders/ipfs-uploader.ts:119]: @iketw I think we should throw at the SDK instantiation - not here

File: packages/wallets/src/evm/connectors/smart-wallet/lib/http-rpc-client.ts [hunk 17]
```diff
@@ -25,12 +25,16 @@ export class HttpRpcClient {
 
     const headers: Record<string, string> = {};
 
-    if (clientId) {
-      headers["x-client-id"] = clientId;
+    if (secretKey && clientId) {
+      throw new Error(
```
Comment (MananTank) [inline packages/wallets/src/evm/connectors/smart-wallet/lib/http-rpc-client.ts:29]: same - throw as soon as possible

File: packages/wallets/src/evm/connectors/smart-wallet/lib/paymaster.ts [hunk 18]
```diff
@@ -32,12 +32,16 @@ class VerifyingPaymasterAPI extends PaymasterAPI {
       "Content-Type": "application/json",
     };
 
-    if (this.clientId) {
-      headers["x-client-id"] = this.clientId;
+    if (this.secretKey && this.clientId) {
+      throw new Error(
+        "Cannot use both secret key and client ID. Please use secretKey for server-side applications and clientId for client-side applications.",
+      );
```
Comment (MananTank) [inline packages/wallets/src/evm/connectors/smart-wallet/lib/paymaster.ts:38]: same - throw as soon as possible

**Comment by iketw** [comment]
./release-pr

**Comment by iketw** [comment]
./release-pr

**Comment by iketw** [comment]
./release-pr

File: packages/sdk/src/evm/core/wallet/user-wallet.ts [hunk 19]
```diff
@@ -294,6 +294,7 @@ export class UserWallet {
       const localWallet = new UserWallet(
         new Wallet(LOCAL_NODE_PKEY, getChainProvider(chainId, this.options)),
         this.options,
+        // this.storage,
```
[the remark this answers is not in the document]
Comment (MananTank) [reply packages/sdk/src/evm/core/wallet/user-wallet.ts:297]: No - forget to delete it
